### PR TITLE
Update to version 3.1.0

### PR DESCRIPTION
**Pull request by jjhelmus**

### CODE REVIEW

**Comment by jjhelmus** [comment]
Not sure if it is safe to remove the patches include in the `3.0.0` but worth a try.

**Comment by jakirkham** [comment]
> Not sure if it is safe to remove the patches include in the `3.0.0` but worth a try.

Yes, should be fine. IIRC both those changes were included in `3.1.0`. So we shouldn't need the patches anymore.

File: recipe/meta.yaml [hunk 1]
```diff
@@ -12,15 +12,9 @@ source:
...
     patches:
-      # Add a patch for non-C++11 support.
-      - field_mask_util.cc.patch
-      # Adds a patch to get rid of some constants also included in a class.
-      - time_util_const_dedup.patch
 
 build:
     number: 0
-    # skip the windows python 2.7 and 3.4 build
-    # see https://github.com/google/protobuf/issues/1958
```
Comment (jakirkham) [inline recipe/meta.yaml]: LOL. Sorry for leaving this errant comment in there. Thanks for cleaning it out.

**Comment by jjhelmus** [comment]
The x64 Python 2.7 windows built is failing when running the tests.  First error seems to be:

```
...
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'><key': true}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'k><ey': true}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'ke><y': true}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key><': true}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key'><: true}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key':>< true}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key': ><true}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key': t><rue}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key': tr><ue}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key': tru><e}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key': true><}}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:111] Testing split: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key': true}><}
[00:16:56] [libprotobuf INFO C:\Miniconda-x64\conda-bld\work\protobuf-3.1.0\src\google\protobuf\util\internal\json_stream_parser_test.cc:97] Testing split every char: {t: true, f: false, n: null, s: 'a string', d: "another string", pi: 22, ni: -127, pd: 45.3, nd: -1056.4, pl: 11779497823553162765, l: [[]], o: {'key': true}}
[00:16:56] unknown file: error: SEH exception with code 0xc00000fd thrown in the test fixture's destructor.
[00:16:56] [  FAILED  ] JsonStreamParserTest.ObjectValues (234 ms)
[00:16:56] [ RUN      ] JsonStreamParserTest.RejectNonUtf8WhenNotCoerced
```

Maybe a different version of googlemock and googletest are needed?

**Comment by jjhelmus** [comment]
> Maybe a different version of googlemock and googletest are needed?

Apparently not.  Will need to investigate further

**Comment by croth1** [comment]
Any progress here? Can we move on by temporarily disabling the failing windows build? This blocks the tensorflow and keras updates.

**Comment by QtRoS** [comment]
@croth1 I agree, current stable release of tensorflow is 12.1, but I still have only 0.11 in conda.

**Comment by babhishek21** [comment]
The Windows build fails at `nmake test` (unit tests). A total of 110 tests are failing. Some people are having troubles with the Unix style line endings for some of the files while being built on Windows (see [here](https://github.com/google/protobuf/issues/2548)). Could this be a possible cause?

**Comment by QtRoS** [comment]
@babhishek21 for me yes, I think it can be the reason of troubles.

**Comment by jjhelmus** [comment]
Re-rendered the feedstock which should trigger a new sets of builds.  Lets see if the Windows builds are still having issues.

**Comment by jakirkham** [comment]
Not that I think it is relevant here, but we also now apply VS 2008 64-bit patch on AppVeyor. Figured I'd mention it in case NMake ends up not being the way to go.

**Comment by jjhelmus** [comment]
Looks like only the Python 2.7 x64 build is failing on Windows.  I can skip this build so that we can get the package out.  Does that sound reasonable?

**Comment by ocefpaf** [comment]
> Does that sound reasonable?

Fine with me. But I am biased b/c I need the Linux package for PY36 out :wink:

**Comment by jjhelmus** [comment]
Added a selector to skip the win-64 python 2.7 build.

**Comment by ocefpaf** [comment]
Thanks @jjhelmus!

**Comment by babhishek21** [comment]
This will help the `keras` and `tensorflow` feedstocks' update. It's kind of ironic given that `tensorflow` finally supports Windows.

**Comment by jjhelmus** [comment]
> It's kind of ironic given that tensorflow finally supports Windows.

Yes but only 64-bit Python 3.5 on Windows, so luckily the skipped protobuf build should not case an issue for `tensorflow`.

**Comment by QtRoS** [comment]
> It's kind of ironic given that tensorflow finally supports Windows.
For me it is ironic that tensorflow moved to v1.0rc while we were updating protobuf :) Just to remind - initial intention was to upgrade tensorflow to v0.12

**Comment by jakirkham** [comment]
Are there any relevant issues upstream or has anyone raised one?

File: recipe/meta.yaml [hunk 2]
```diff
@@ -12,19 +12,15 @@ source:
...
     features:
         - vc9                # [win and py27]
         - vc10               # [win and py34]
         - vc14               # [win and py35]
```
Comment (jakirkham) [inline recipe/meta.yaml]: I think we need a feature for Python 3.6, correct?

Comment (jjhelmus) [inline recipe/meta.yaml]: Agreed, added in 270e1b2

**Comment by jjhelmus** [comment]
I found a number of reports of failing unit tests on Windows with protobuf 3.1.0, google/protobuf#2548, google/protobuf#2238 and google/protobuf#2408.  It has been suggested that this may be cause by a "\r\n" vs. "\n" problem and the failures can be ignored.  

I do not have the time to track these down or investigate the issue more. If some else would like to take this PR over I'd be happy to let them.

**Comment by ocefpaf** [comment]
> So I'd be ok with skipping Python 2.7 64-bit and merging an otherwise passing PR given the following constraints are met.
> 1. An issue is raised upstream with logs and info about the failures encountered here.
> 2. An issue is raised on this feedstock about getting Python 2.7 64-bit back up and running referencing the upstream issue.
> Once these constraints are met I would be happy to approve this.

This is too much for the maintainer. I only an issue here is enough. Everybody is quite busy and people should not be denied of working packages because of that.

**Comment by jakirkham** [comment]
> I do not have the time to track these down or investigate the issue more.

Certainly, nor was this a request to do so. Just trying to collect the information that you have so someone can address it when they are able. [Your comment]( https://github.com/conda-forge/protobuf-feedstock/pull/16#issuecomment-272472630 ) above was quite helpful in that regard.

**Comment by dopplershift** [comment]
You may not have intended it, but your comment came across as a demand, not a friendly request. Requests are asked--demands are stated.

**Comment by babhishek21** [comment]
@dopplershift I think we are not giving credit to @jakirkham. He was assigned as a reviewer, right? Well, this PR does have some outstanding issues attached to it. He is doing the right thing by ensuring that those are not forgotten. Maintainers have a commitment to the community to *maintain* systems. He is just trying to do his job. 

That said, @jakirkham, your comment may have come across as too strong a request. Next time, maybe "suggest." I am sure the community would rationalise on what's best.

**Comment by dopplershift** [comment]
Yes, but @ocefpaf, @jjhelmus, @jakirkham, and myself are **all** maintainers.

And really, I'm not faulting @jakirkham so much as based on the "nor was it a request to do so", I felt he was missing how forceful he came across, based on my reactions of all of the other *maintainers* of this recipe.

**Comment by jakirkham** [comment]
Thanks for the feedback. I'll keep that in mind for the future.

Was not trying to hold up progress, make undue requests, or come across as forceful. Am well aware of the difficulties building for VS 2008 64-bit. Particularly because of the time I and others have spent debugging them. Was merely trying to convey what I would believe to be standard practice in this case and make sure it didn't get lost in the already quite long thread. Namely that we document the issue we ran into and get upstream's attention on it. This all in hopes of making it easier for the next person who tries to address this problem.

**Pull request merged**